contracts/facets/YGFacetZaynFi.sol

```diff
@@ -119,4 +122,110 @@ contract YGFacetZaynFi is IYGFacetZaynFi {
+    /// @param termIds The term ids for which the yield is being rescued
+    /// @param originalWithdrawals The original ETH withdrawal amounts of each bad transaction
+    /// @param originalShares The original calculated shares amounts of each bad transaction
+    /// @param users The users to be reimbursed
+    function rescueStuckYields(uint[] memory termIds, uint[] memory originalWithdrawals, uint[] memory originalShares, address[] memory users) external onlyOwner payable {
+        // Start with validating the lengths of the arrays
+        uint length = termIds.length;
+        require(length == originalWithdrawals.length && length == originalShares.length && length == users.length, "Arrays don't match");
+
+        uint usedValue = 0; // Used to keep track of the lost ETH stored back into zaynfi
+
+        // Start looping through each combination
+        for (uint i; i < length; ) {
+            uint termId = termIds[i];
+            address user = users[i];
+
+            LibYieldGenerationStorage.YieldGeneration storage yield = LibYieldGenerationStorage
+                ._yieldStorage()
+                .yields[termId];
+
+            // Make sure user is part of this term and has enabled yield generation
+            require(yield.hasOptedIn[user], "User not part of yield generation");
+
+            // Zaynfi's addresses
+            address vaultAddress = yield.providerAddresses["ZaynVault"];
+            address zapAddress = yield.providerAddresses["ZaynZap"];
+
+            // Calculate what each user is owed
+            int reimbursement = _calculateReimbursement(originalWithdrawals[i], originalShares[i], yield);
+
+            if (reimbursement > 0) {
+                // Reimbursement is positive, this means the user withdrew less shares than he was supposed to
+                uint neededShares = uint(reimbursement);
+
+                // Code copied from _withdrawYG, get the amount of shares back and give it to the user
+                uint withdrawnYield = IZaynZapV2TakaDAO(zapAddress).zapOutETH(
+                    vaultAddress,
+                    neededShares,
+                    termId
+                );
+
+                yield.withdrawnYield[user] += withdrawnYield;
+                yield.availableYield[user] += withdrawnYield;
+
+                // Claim the yield right away and send it to the user
+                LibYieldGeneration._claimAvailableYield(termId, user, user);
+                
+            } else if (reimbursement < 0) {
+                // When there is a negative reimbursement, we compensate the pool by adding back the exact amount of shares that were lost
+                uint neededShares = uint(reimbursement * -1);
+
+                // Calculate the amount of eth we need to deposit to get the desired shares
+                uint pricePerShare = IZaynVaultV2TakaDao(vaultAddress)
+                    .getPricePerFullShare();
+
+                uint neededEth = neededShares * pricePerShare;
```

Comment:
Here I fixed the decimals with neededShares * pricePerShare / 10 ** 18

Reply:
Yeah forgot that, my bad. Please commit